hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientMapProxy.java

```diff
@@ -1670,6 +1673,67 @@ private void putAllInternal(@Nonnull Map<? extends K, ? extends V> map,
+        ClientPartitionService partitionService = getContext().getPartitionService();
+
+        Map<Integer, List<SimpleEntryView<Data, Data>>> entriesByPartition =
+                entries.stream()
+                       .map(e -> {
+                           checkNotNull(e.getKey(), NULL_KEY_IS_NOT_ALLOWED);
+                           checkNotNull(e.getValue(), NULL_VALUE_IS_NOT_ALLOWED);
+
+                           Data keyData = toData(e.getKey());
+                           if (e instanceof SimpleEntryView
+                                   && e.getKey() instanceof Data
+                                   && e.getValue() instanceof Data
+                           ) {
+                               return (SimpleEntryView<Data, Data>) e;
+                           } else {
+                               return new SimpleEntryView<>(keyData, toData(e.getValue()))
+                                       .withCost(e.getCost())
+                                       .withCreationTime(e.getCreationTime())
+                                       .withExpirationTime(e.getExpirationTime())
+                                       .withHits(e.getHits())
+                                       .withLastAccessTime(e.getLastAccessTime())
+                                       .withLastStoredTime(e.getLastStoredTime())
+                                       .withLastUpdateTime(e.getLastUpdateTime())
+                                       .withVersion(e.getVersion())
+                                       .withTtl(e.getTtl())
+                                       .withMaxIdle(e.getMaxIdle());
+                           }
+                       })
+                       .collect(groupingBy(
+                               (SimpleEntryView<Data, Data> e) -> partitionService.getPartitionId(e.getKey())
+                       ));
+
+        AtomicInteger counter = new AtomicInteger(entriesByPartition.size());
+        List<CompletableFuture<Void>> futures = new ArrayList<>(entriesByPartition.size());
+        for (Entry<Integer, ? extends List<SimpleEntryView<Data, Data>>> entry : entriesByPartition.entrySet()) {
+            Integer partitionId = entry.getKey();
+            ClientMessage request = MapPutAllWithMetadataCodec.encodeRequest(name, entry.getValue());
+            ClientInvocationFuture future = new ClientInvocation(getClient(), request, getName(), partitionId)
+                    .invoke();
+
+            future.whenCompleteAsync((clientMessage, throwable) -> {
+                        if (counter.decrementAndGet() == 0) {
+                            finalizePutAll(
+                                    entries,
+                                    entriesByPartition
+                            );
+                        }
+                    }, ConcurrencyUtil.getDefaultAsyncExecutor());
+            futures.add(new ClientDelegatingFuture<>(future, getSerializationService(), clientMessage -> null));
```

Comment:
nit: you could avoid creating additional `ClientDelegatingFuture` instances by creating & returning a single `InternalCompletableFuture` instance which is completed in the `whenCompleteAsync` callback with `null` instead of returning `CompletableFuture.allOf` (simlarly to `putAllInternal` implementation).
The bonus for returning `InternalCompletableFuture` is that you can use `joinInternal()` from the calling side, which takes care of some exception un-/rewrapping according to the usual calling convention of our public APIs. But I don't know if that's relevant for you, so feel free to keep the existing implementation.

Reply:
I think I managed to do this as well.